ts/packages/ts/src/commands/UpsertCommand.ts

```diff
@@ -5,59 +5,94 @@ import { ConfigError } from '../utils';
 
-    await store.init();
+    const errorScope: [string, string][] = [
+      ['UpsertCommand', `store:${store.type};set:${set.path};schema:${schema.name}`],
+      ['parameters.configs', `key:${key};value:${value}`],
+    ];
 
-    const upsertConfigs = _(configs)
-      .entries()
-      .map<Config>(([key, value]) => {
-        const errorScope: [string, string][] = [
-          ['UpsertCommand', `store:${store.type};set:${set.path};schema:${schema.name}`],
-          ['parameters.configs', `key:${key};value:${value}`],
-        ];
+    const cfgu = schema.contents[key];
 
-        const cfgu = schema.contents[key];
+    if (!skipDeclarationAndTemplateValidation) {
+      if (!cfgu) {
+        throw new ConfigError(
+          'invalid config key',
+          `key "${key}" must be declared on schema ${schema.name}`,
+          errorScope,
+        );
+      }
+      if (value !== undefined && cfgu?.template) {
+        throw new ConfigError('invalid config value', `keys declared with template mustn't have a value`, errorScope);
+      }
+    }
 
-        if (!cfgu) {
-          throw new ConfigError(
-            'invalid config key',
-            `key "${key}" must be declared on schema ${schema.name}`,
-            errorScope,
-          );
-        }
+    if (value !== undefined && cfgu) {
+      try {
+        ConfigSchema.CFGU.VALIDATORS.valueOptions(cfgu, value);
+        ConfigSchema.CFGU.VALIDATORS.valueType(cfgu, value);
+      } catch (error) {
+        throw error?.appendScope?.(errorScope) ?? error;
+      }
+    }
+  }
+
+  async run() {
+    const { store, set, schema, configs, pipe } = this.parameters;
 
-        if (value) {
-          if (cfgu.template) {
-            throw new ConfigError(
-              'invalid config value',
-              `keys declared with template mustn't have a value`,
-              errorScope,
-            );
-          }
+    if (_.isEmpty(configs) && _.isEmpty(pipe)) {
+      return;
+    }
 
-          try {
-            ConfigSchema.CFGU.VALIDATORS.valueOptions(cfgu, value);
-            ConfigSchema.CFGU.VALIDATORS.valueType(cfgu, value);
-          } catch (error) {
-            throw error?.appendScope?.(errorScope) ?? error;
-          }
+    await store.init();
+
+    const pipeConfigs = _(pipe)
+      .entries()
+      .filter(([key]) => {
+        const schemaCfgu = schema.contents[key];
+        if (!schemaCfgu) {
+          return false;
         }
+        return !schemaCfgu.template;
+      })
+      .map(([key, value]) => {
+        this.validateConfigValue({ key, value: value.result.value, skipDeclarationAndTemplateValidation: true });
+        return {
+          set: set.path,
+          key,
+          value: value.result.value,
+        };
+      })
+      .value();
 
+    const upsertConfigs = _(configs)
+      .entries()
+      .map<Config>(([key, value]) => {
+        this.validateConfigValue({ key, value, skipDeclarationAndTemplateValidation: false });
```

Comment:
I would convert the whole mapper function here

Reply:
split the skip flag to each of the validation

Reply:
I split the validations. I don't understand what "convert the whole mapper function here" means.. please elaborate.